Update base path for Beats installation on Windows

belimawr opened the pull request:
## Proposed commit message

The base folder for data and logs on Windows has changed from
`C:\ProgramData\` to `C:\Program Files\` because the latter has more
strict permissions.

When the installation script runs, it looks for the previous default
data path, if it is found, the data is moved to the new path.
The installation script accepts the parameter `-UseLegacyPath` to
force using the legacy data path.

The old paths were:
- Installation path: `C:\Program Files\Filebeat`
- Home path: `C:\Program Files\Filebeat`
- Data path: `"C:\ProgramData\filebeat`
- Logs path: `"C:\ProgramData\Filebeat\logs`

The new paths are:
- Installation path: `C:\Program Files\Filebeat`
- Home path: `C:\Program Files\Filebeat`
- Data path: `"C:\Program Files\Filebeat-Data`
- Logs path: `"C:\Program Files\Filebeat-Data\logs`

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] ~~I have made corresponding change to the default configuration files~~
- [ ] ~~I have added tests that prove my fix is effective or that my feature works~~
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Disruptive User Impact
This PR updates the default data path for Beats when installed on Windows and moves any existing data during an upgrade to the new path. This behaviour can be overridden by providing the CLI argument `-UseLegacyPath`

## Author's Checklist
- [x] Verify if we should move the data folder from the old default to the new default
- [x] Decide whether to keep the `Write-Debug` statements in the install script.
- [x] Move the data path to `C:\Program Files`

## How to test this PR locally

### The normal installation

1. Package Filebeat for Windows:
   ```
   DEV=true SNAPSHOT=true PACKAGES="zip" PLATFORMS=windows/amd64 mage -v package
   ```
2. Follow [the instructions](https://www.elastic.co/docs/reference/beats/filebeat/filebeat-installation-configuration#installation) to install Filebeat on Windows
3. Ensure the Windows service is installed having `C:\Program Files\` as the base path for `--path.home`, `--path.data` and `--path.logs`. You can get the full command line for the service by running (on Powershell):
   ```
   (Get-CimInstance Win32_Service -Filter 'Name = "filebeat"').PathName
   ```

### Upgrading from an older version
1. Download any release version of Filebeat, the latest is always the recommended one
2. Follow [the instructions](https://www.elastic.co/docs/reference/beats/filebeat/filebeat-installation-configuration#installation) to install Filebeat on Windows
3. Ensure the Windows service is installed having `C:\ProgramData\` (actually it's the vaule from `$env:ProgramData`) as the base path for `--path.home`, `--path.data` and `--path.logs`. You can get the full command line for the service by running (on Powershell):
   ```
   (Get-CimInstance Win32_Service -Filter 'Name = "filebeat"').PathName
   ```
4. Package Filebeat for Windows:
   ```
   DEV=true SNAPSHOT=true PACKAGES="zip" PLATFORMS=windows/amd64 mage -v package
   ```
5. Stop the Filebeat service:
   ```
   (Get-Service filebeat).Stop()
   ```
6. Follow the instructions from `#2` to upgrade Filebeat, make sure to run the new `install-service-filebeat.ps1` script
7. Ensure the Windows service is installed having `C:\Program Files\` (actually it's the value from `$env:ProgramFiles`) as the base path for `--path.home`, `--path.data` and `--path.logs` and the previous data folder is moved to `C:\Program Files\`. You can get the full command line for the service by running, (on Powershell):
   ```
   (Get-CimInstance Win32_Service -Filter 'Name = "filebeat"').PathName
   ```


### Bonus test
1. Package Filebeat for Windows:
   ```
   DEV=true SNAPSHOT=true PACKAGES="zip" PLATFORMS=windows/amd64 mage -v package
   ```
2. Follow [the instructions](https://www.elastic.co/docs/reference/beats/filebeat/filebeat-installation-configuration#installation) to install Filebeat on Windows, but stop before running the `install-service-filebeat.ps1` script.
3. Run:
   ```
   Get-Help .\install-service-filebeat.ps1 -detailed
   ```
4. Read the docs and test the supported parameters

~~## Related issues~~
~~## Use cases~~
~~## Screenshots~~
~~## Logs~~

belimawr commented:
I created a flaky test issue for the test that failed: https://github.com/elastic/beats/issues/44032

elasticmachine commented:
Pinging @elastic/elastic-agent-data-plane (Team:Elastic-Agent-Data-Plane)

leehinman commented on dev-tools/packaging/templates/windows/install-service.ps1.tmpl:
```diff
@@ -6,12 +38,37 @@ if (Get-Service {{.BeatName}} -ErrorAction SilentlyContinue) {
+# We need to support a new default path for the data folder, ideally
+# automatically detecting if the old one is used and keeping it
+
 $workdir = Split-Path $MyInvocation.MyCommand.Path
```
Nit.  Any chance we could change from `workdir` to `WorkDir` so it matches all the other variables (HomePath, LogsPath, etc)?

belimawr replied:
Done: https://github.com/elastic/beats/pull/43995/commits/5cd87f5142ddc6820b4596fe5f58cfbc527a11db

leehinman reviewed (approved):
LGTM, one non blocking nit below.

strawgate commented:
Please hold off on merging this for now, will follow-up soon

jlind23 commented:
@belimawr after chatting with @strawgate and if our understanding is correct, already existing users that will not supply `ForceNewDataPath` will keep on relying on their old ProgramData path right? Can't we change this default instead?

belimawr commented:
> @belimawr after chatting with @strawgate and if our understanding is correct, already existing users that will not supply `ForceNewDataPath` will keep on relying on their old ProgramData path right? 

Yes, that is correct. If just using the script to upgreade/re-install, the current data path is kept, so the states are kept.


> Can't we change this default instead? I believe we should keep on "forcing" users to use `Program Files` unless they explicitely specify `UseLegacyPath` thoughts?

Yes, we can. I added some thoughts about it in our internal issue and got no comments, so I just moved on with the PR.

Anyways, I'll replicate them here.

> Another feature we could add is moving the data folder during the installation from C:\ProgramData to C:\Program Files, either by default or as an opt-in. However, doing this automatically can cause issues if the users rely on the data path for things like shipping logs to ES or even backups, which could be a pretty annoying breaking change.

So, yes it is totally feasible to transparently move the data path to `C:\Program Files` keeping the registry. I do not foresee any issues because Filebeat (and the install script) already runs with elevated privileges, so it should be no issue moving the data folder from `C:\ProgramData` to `C:\Program Files`.

What I'd do in any error while copying the files it to stop the script, with clear messages indicating what we're doing and how to do it manually so users can fix their deployment and not have data duplication.

strawgate commented:
We should make sure that whatever we do to copy the files does copy permissions

belimawr commented:
I'm moving this back to draft until I finish implementing the new logic @jlind23 asked for https://github.com/elastic/beats/pull/43995#issuecomment-2827613581.

belimawr commented:
@leehinman, @khushijain21 the PR is ready for review again. I updated it to actually move the data path to `C:\Program Files\<Beat name>-data`.

cmacknz commented:
How do we test this? Can we automate it? Or give QA a description for a manual test?

Also you are missing a changelog entry.

cmacknz commented on docs/reference/auditbeat/auditbeat-installation-configuration.md:
```diff
@@ -253,6 +253,10 @@ Auditbeat comes with predefined assets for parsing, indexing, and visualizing yo
 
+
+By default, Windows log files are stored in `C:\Program Files\auditbeat-Data\logs`.
+For versions < 9.1.0, logs are stored by default in `C:\ProgramData\auditbeat\Logs`.
```
This isn't right, the log path you use is `$LogsPath = "$HomePath\log"` which doesn't match these docs.

belimawr replied:
That was a typo in the script, I've fixed it already. Thanks for catching it!

cmacknz commented on docs/release-notes/breaking-changes.md:
```diff
@@ -12,6 +12,20 @@ Breaking changes can impact your Elastic applications, potentially disrupting no
 % Description and impact of the breaking change.
 % For more information, check [PR #](PR link).
 
+::::{dropdown} The default data and logs path for the Windows service installation has changed.
```
You should mention what the new directory structure is

cmacknz commented:
Include an example of the before and after directory structures. It is weirdly inconsistent I think, some of the directory names have changed unnecessarily.

belimawr commented:
> How do we test this? (...) Or give QA a description for a manual test?

Currently, manual testing, the procedure is in the "How to test this PR locally" section of this PR description.

> Can we automate it?

It is possible, however it is going to be a big endeavour because we have no testing infrastructure on Beats to run tests within VMs and our integration tests do not run on Windows. We need something like the the Elastic-Agent has for coordinating VMs to run the tests because we will be installing Filebeat in the default path and creating a system service, this requires being sure there is no other Filebeat running (like for monitoring of the host) and a good clean up procedure, which is much easier on a ephemeral VM.

belimawr commented on dev-tools/packaging/templates/windows/install-service.ps1.tmpl:
```diff
@@ -6,12 +33,45 @@ if (Get-Service {{.BeatName}} -ErrorAction SilentlyContinue) {
+}
+
+$HomePath = "$BasePath\{{.BeatName}}"
+$LogsPath = "$HomePath\log"
```
Yes, I didn't mean to change it, it was a typo :man_facepalming:

belimawr commented on dev-tools/packaging/templates/windows/install-service.ps1.tmpl:
```diff
@@ -1,3 +1,30 @@
+
+    You can pass ForceLegacyPath to use the legacy data path.
+
+    If the Windows service already exits, it will be stopped and deleted, then
```
Thanks, I'll add that together with the case change
